core/src/main/java/org/bitcoinj/wallet/Wallet.java

```diff
@@ -288,7 +288,14 @@ protected boolean removeEldestEntry(Map.Entry<Sha256Hash, Transaction> eldest) {
     protected volatile TransactionBroadcaster vTransactionBroadcaster;
     // UNIX time in seconds. Money controlled by keys created before this time will be automatically respent to a key
     // that was created after it. Useful when you believe some keys have been compromised.
-    private volatile long vKeyRotationTimestamp;
```

Comment:
Note that in the description you say `MAX`, here you implement `MIN`. `MIN` is closer to the previous `0`, however if the special case isn't handled correctly it could lead into a mass-rotation of wallets. `MAX` somehow protects against that I'd say.

Either way, I don't really like it. `MAX` or `MIN` are magic constants. How is it better than using `null`?

Looking at it at a broader scale, magic constants like those are not always available. `null` or `Optional` can be used with any type (also native types using autoboxing).

Reply:
> Note that in the description you say `MAX`, here you implement `MIN`. `MIN` is closer to the previous `0`, however if the special case isn't handled correctly it could lead into a mass-rotation of wallets. `MAX` somehow protects against that I'd say.

Yes, I've been a little bit confused and kept changing my WIP code. There are different use-cases and different meanings it seems for some of the uses of 0.  (e.g. in a key creation time, it means unknown but for key rotation 0 has a different meaning)

> 
> Either way, I don't really like it. `MAX` or `MIN` are magic constants. How is it better than using `null`?

I think `null` is probably fine (and possibly "better") in this use case. But, I do think that sometimes it is better to use 0, MAX, or MIN instead of null -- especially in cases where a "natural" integer comparison will handle the special case without having to worry about NPE.

(As I said elsewhere my opinion has changed on this in recent years as I've been adopting "functional-style" Java. My abhorrence for `null` is now greater than my abhorrence for "magic constants".)

> 
> Looking at it at a broader scale, magic constants like those are not always available. `null` or `Optional` can be used with any type (also native types when using autoboxing).

Yes. And I think (as you've commented elsewhere) in the general case `Optional` is preferred for return types and for parameters it is generally preferred to use either nullable or overloading.

Reply:
>

Reply:
> Also I'm specifically interested in how you would refactor the `getKeyRotationTime()` accessor. I assume it would get a `getKeyRotationTimeInstant()` sibling, but would you expose the magic constant here, too? (This is an even more important question that the above, as it touched public API.)

In my work last night, I was unable to satisfactorily answer this question. Part of the problem (IMHO) is that I didn't see any comments that clearly explained what these timestamps meant (I may not have looked carefully enough.) So I don't really have an answer yet.